### PR TITLE
Improve Canada Payroll

### PR DESCRIPTION
**Pull request by ghost**
This module implements the canadian payroll structure. In order to pass the travis tests, the following pull requests need to pass:
~~https://github.com/OCA/l10n-canada/pull/36~~
~~https://github.com/OCA/l10n-canada/pull/32~~

### CODE REVIEW

**Comment by max3903** [comment]
Both PR have been merged.

**Comment by ghost** [comment]
This pull request needs to be merged too https://github.com/OCA/hr/pull/58.

**Comment by ghost** [comment]
This pull request needs to be merged too hr_contract_hourly_rate.

**Comment by ghost** [comment]
https://github.com/OCA/hr/pull/53

**Comment by bwrsandman** [comment]
Changes need to be rebased

**Comment by ghost** [comment]
PR https://github.com/OCA/hr/pull/40 needs to pass first.

File: l10n_ca_hr_payroll/README.rst [hunk 1]
```diff
@@ -0,0 +1,57 @@
...
+=======
+
+Contributors
+------------
+
+.. image:: http://sflx.ca/logo
+   :alt: Savoir-faire Linux
+   :target: http://sflx.ca
+
+* David Dufresne <david.dufresne@savoirfairelinux.com>
```
Comment (max3903) [inline l10n_ca_hr_payroll/README.rst]: Please add Amura Consulting like in the AUTHORS file

File: l10n_ca_hr_payroll/hr_cra_t4_transmission.py [hunk 2]
```diff
@@ -0,0 +1,270 @@
...
+#    MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
+#    GNU Affero General Public License for more details.
+#
+#    You should have received a copy of the GNU Affero General Public License
+#    along with this program.  If not, see <http://www.gnu.org/licenses/>.
+#
+##############################################################################
+
+from openerp.osv import fields, orm
+from dict2xml import dict2xml
```
Comment (max3903) [inline l10n_ca_hr_payroll/hr_cra_t4_transmission.py]: This dependency is not listed in the external_dependencies of `__openerp__.py`

Why don't you use lxml ? It is already a dependency of Odoo.

Comment (ghost) [inline l10n_ca_hr_payroll/hr_cra_t4_transmission.py]: Didn't know lxml until yesterday. I will push the change soon.

**Comment by bwrsandman** [comment]
@dufresnedavid could you squash some of your commits? Especially commits such as Fix flake8.
76 commits is very excessive tor review for one pull request.

Also, remember that it is suggested to use the present participle when writing commit messages. As in this commit does what vs what did this commit do.
- Yes: Change nas to sin
- No: Fixed T4 report

**Comment by ghost** [comment]
@bwrsandman Thanks, I will squash commits as soon as I can.

**Comment by ghost** [comment]
~~https://github.com/OCA/hr/pull/72 needs to be merged first~~

File: l10n_ca_hr_payroll/hr_cra_transmission.py [hunk 3]
```diff
@@ -0,0 +1,325 @@
...
+#    You should have received a copy of the GNU Affero General Public License
+#    along with this program.  If not, see <http://www.gnu.org/licenses/>.
+#
+##############################################################################
+
+from openerp.osv import orm, fields
+from openerp.tools.translate import _
+from openerp.tools import DEFAULT_SERVER_DATE_FORMAT
+from lxml import etree
+import iso3166
```
Comment (max3903) [inline l10n_ca_hr_payroll/hr_cra_transmission.py]: Please add this lib in your external dependencies in `__openerp__.py`

**Comment by ghost** [comment]
@max3903 The csv files are encoded and cannot be opened with Libre Office.

**Comment by bwrsandman** [comment]
Depends on 
- [x] https://github.com/OCA/hr/pull/104
- [x] https://github.com/OCA/hr/pull/82

**Comment by ghost** [comment]
@bwrsandman @max3903 I didn't expect all 10 of my PR in OCA/hr to be merged in one day. 
I am preparing an important commit for l10n_ca_hr_payroll and I need about a week.

**Comment by ghost** [comment]
@bwrsandman You don't need to review the files in l10n_ca_hr_payroll/data/cra_xml_schemas. It is the official xml specifications from Canada Revenu. I only use it for unit test purposes.

**Comment by bwrsandman** [comment]
@dufresnedavid Do you want me to label this as a work in progress?

**Comment by ghost** [comment]
@bwrsandman Yes please

**Comment by ghost** [comment]
https://github.com/OCA/l10n-canada/pull/68 needs to pass first

**Comment by ghost** [comment]
@bwrsandman I have a last PR at ~~https://github.com/OCA/hr/pull/113~~ that needs to pass before this one.

File: l10n_ca_hr_income_tax_deductions/tests/test_deduction.py [hunk 4]
```diff
@@ -0,0 +1,187 @@
...
+
+        payslip = self.payslip_model.browse(
+            cr, uid, self.payslip_id, context=context)
+
+        return {
+            line.code: line.total
+            for line in payslip.details_by_salary_rule_category
+        }
+
+    def test_sum_deductions(self):
```
Comment (bwrsandman) [inline l10n_ca_hr_income_tax_deductions/tests/test_deduction.py:170]: This test is failing

Comment (ghost) [inline l10n_ca_hr_income_tax_deductions/tests/test_deduction.py:170]: Fixed

**Comment by ghost** [comment]
@bwrsandman you may remove the work in progress label

**Comment by ghost** [comment]
@max3903 @bwrsandman Travis is green. Runbot is red because of account_statement_td_canada module. Sandy, can you had the missing dependency of the td module ?

**Comment by max3903** [comment]
@dufresnedavid Missing dependency added.

**Comment by ghost** [comment]
@max3903 missing num2words in runbot deps. Can you also check for the python library iso3166? I think it is missing too.

**Comment by max3903** [comment]
@dufresnedavid I just installed num2words and iso3166 on runbots.

**Comment by ghost** [comment]
@max3903 the module l10n_ca_account_profile has missing dependency currency_rate_update.

**Comment by max3903** [comment]
@dufresnedavid Should be solved now.

**Comment by ghost** [comment]
@max3903 There are still missing dependencies for runbot.

**Comment by max3903** [comment]
@dufresnedavid I added reporting-engine repo.

**Comment by ghost** [comment]
@max3903 runbot is missing OCA/account-invoice-reporting

**Comment by max3903** [comment]
@dufresnedavid Added! Here is the current list of repo:

github.com/OCA/OCB 
github.com/OCA/account-financial-reporting
github.com/OCA/account-financial-tools 
github.com/OCA/account-fiscal-rule
github.com/OCA/account-invoice-reporting 
github.com/OCA/account-invoicing 
github.com/OCA/account-payment 
github.com/OCA/bank-statement-reconcile 
github.com/OCA/hr 
github.com/OCA/reporting-engine
github.com/OCA/webkit-tools 
github.com/OCA/bank-payment

**Comment by ghost** [comment]
@max3903 I still get this error from runbot:
except_orm: (u'Error', u"You try to install module 'l10n_ca_account_profile' that depends on module 'partner_aging'.\nBut the latter module is not available in your system.")

**Comment by max3903** [comment]
- [x] https://github.com/OCA/account-invoice-reporting/pull/32

File: l10n_ca_hr_payroll/tests/test_payroll_structure_leaves.py [hunk 5]
```diff
@@ -0,0 +1,406 @@
...
+            round(payslip_2['VAC_AVAIL_HOURS'], 2),
+            round(40 - 10 + 160 * 0.08, 2))
+
+        self.assertEqual(
+            round(payslip_2['VAC_REQ'], 2),
+            round((4 * 8 + 20) * 40, 2))
+
+        self.assertEqual(payslip_2['VAC_TAKEN'], payslip_2['VAC_AVAIL'])
+        self.assertEqual(
+            payslip_2['VAC_TAKEN_HOURS'], payslip_2['VAC_AVAIL_HOURS'])
```
Comment (bwrsandman) [inline l10n_ca_hr_payroll/tests/test_payroll_structure_leaves.py:215]: This test fails

**Pull request closed without merging**